components/org.wso2.carbon.identity.workflow.engine/src/main/java/org/wso2/carbon/identity/workflow/engine/ApprovalTaskServiceImpl.java

```diff
@@ -0,0 +1,631 @@
+import org.wso2.carbon.identity.workflow.engine.model.TaskModel;
+import org.wso2.carbon.identity.workflow.engine.model.TaskParam;
+import org.wso2.carbon.identity.workflow.engine.util.WorkflowEngineConstants;
+import org.wso2.carbon.identity.workflow.mgt.WorkflowManagementService;
+import org.wso2.carbon.identity.workflow.mgt.WorkflowManagementServiceImpl;
+import org.wso2.carbon.identity.workflow.mgt.bean.Parameter;
+import org.wso2.carbon.identity.workflow.mgt.bean.RequestParameter;
+import org.wso2.carbon.identity.workflow.mgt.bean.Workflow;
+import org.wso2.carbon.identity.workflow.mgt.bean.WorkflowAssociation;
+import org.wso2.carbon.identity.workflow.mgt.callback.WSWorkflowCallBackService;
+import org.wso2.carbon.identity.workflow.mgt.callback.WSWorkflowResponse;
+import org.wso2.carbon.identity.workflow.mgt.dao.WorkflowRequestAssociationDAO;
+import org.wso2.carbon.identity.workflow.mgt.dto.Association;
+import org.wso2.carbon.identity.workflow.mgt.dto.WorkflowRequest;
+import org.wso2.carbon.identity.workflow.mgt.exception.InternalWorkflowException;
+import org.wso2.carbon.identity.workflow.mgt.exception.WorkflowClientException;
+import org.wso2.carbon.identity.workflow.mgt.exception.WorkflowException;
+
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+import static org.wso2.carbon.identity.workflow.engine.util.WorkflowEngineConstants.ParameterName.ENTITY_TYPE_CLAIMED_USERS;
+import static org.wso2.carbon.identity.workflow.engine.util.WorkflowEngineConstants.ParameterName.ENTITY_TYPE_ROLES;
+import static org.wso2.carbon.identity.workflow.engine.util.WorkflowEngineConstants.ParameterName.ENTITY_TYPE_USERS;
+
+/**
+ * Implementation of the ApprovalTaskService interface.
+ */
+public class ApprovalTaskServiceImpl implements ApprovalTaskService {
+
+     private static final String APPROVED = "APPROVED";
+    private static final String REJECTED = "REJECTED";
+    private static final Integer LIMIT = 20;
+    private static final Integer OFFSET = 0;
+    private static final Logger log = LoggerFactory.getLogger(ApprovalTaskServiceImpl.class);
+    protected long localCreatedTime;
+
+    private final ApprovalTaskDAO approvalTaskDAO = new ApprovalTaskDAOImpl();
+    private final WorkflowRequestDAO workflowRequestDAO = new WorkflowRequestDAOImpl();
+    private final WSWorkflowCallBackService wsWorkflowCallBackService = new WSWorkflowCallBackService();
+
+    @Override
+    public List<ApprovalTaskSummaryDTO> listApprovalTasks(Integer limit, Integer offset, List<String> status)
+            throws WorkflowEngineException {
+
+        String userId = CarbonContext.getThreadLocalCarbonContext().getUserId();
+
+        List<ApprovalTaskSummaryDTO> approvalTaskSummaryDTOS = getAllAssignedTasks(status, userId);
```

Comment:
Try to improve getAllAssignedTasks to remove duplicated workflow requests, preserve reserved tasks etc..